File: 0024-swap-nodes-in-pairs/24.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *swapPairs(struct ListNode *head) {
  /* Approach:
     1. [Recursion] -- T.C: O(N), S.C: O(N)
     2. [Iterative] -- T.C: O(N), S.C: O(1)
  */

  // Iterative
  struct ListNode *dummy = (struct ListNode *)malloc(sizeof(struct ListNode));
  dummy->val             = 0;
  dummy->next            = head;
  struct ListNode *prev  = dummy;

  while (prev->next && prev->next->next) {
    struct ListNode *first  = prev->next;
    struct ListNode *second = prev->next->next;

    prev->next   = second;
    first->next  = second->next;
    second->next = first;

    prev = first;
  }
  struct ListNode *new_head = dummy->next;
  free(dummy);
  return new_head;

  /* Recursion
  if (head == NULL || head->next == NULL)
    return head;

  struct ListNode *new_head = head->next;
  head->next                = swapPairs(new_head->next);
  new_head->next            = head;
  return new_head;
  */
}

File: 0024-swap-nodes-in-pairs/24.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = Some(Box::new(ListNode {val: 0, next: head}));
        let mut prev = &mut dummy;

        while let Some(ref mut node1) = prev.as_mut().unwrap().next {
            if let Some(mut node2) = node1.next.take() {
                node1.next = node2.next.take();
                node2.next = Some(node1.clone());
                prev.as_mut().unwrap().next = Some(node2);

                prev = &mut prev.as_mut().unwrap().next.as_mut().unwrap().next;
            } else {
                break;
            }
        }

        dummy.unwrap().next
    }
}
